src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java

```diff
@@ -692,93 +690,124 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
+
+									try {
+										sendAndReceive(msg4Confirm, false); //do not follow redirects
+									} catch (SocketException ex) {
+										if (log.isDebugEnabled()) log.debug("Caught " + ex.getClass().getName() + " " + ex.getMessage() + 
+												" when accessing: " + msg4Confirm.getRequestHeader().getURI().toString());
+										continue; //Something went wrong, continue to the next item in the loop
+									}
+									countExpressionBasedRequests++;
+
+									String confirmExpressionOutputUnstripped = msg4Confirm.getResponseBody().toString();
+									String confirmExpressionOutputStripped = this.stripOff(confirmExpressionOutputUnstripped, modifiedParamValueConfirm);
+
+									//set up two little arrays to ease the work of checking the unstripped output or the stripped output
+									String confirmExpressionBodyOutput[] = {confirmExpressionOutputUnstripped, confirmExpressionOutputStripped};
+
+									if (confirmExpressionBodyOutput[booleanStrippedUnstrippedIndex].compareTo(normalBodyOutput[booleanStrippedUnstrippedIndex]) != 0) {
+										//the confirm query did not return the same results.  This means that arbitrary queries are not all producing the same page output.
+										//this means the fact we earier reproduced the original page output with a modified parameter was not a coincidence
+
+										//Likely a SQL Injection. Raise it
+										String extraInfo = null;
+										if (strippedOutput[booleanStrippedUnstrippedIndex]) {
+											extraInfo = Constant.messages.getString(MESSAGE_PREFIX + "alert.expressionbased.extrainfo", modifiedParamValue, "");
+										} else {
+											extraInfo = Constant.messages.getString(MESSAGE_PREFIX + "alert.expressionbased.extrainfo", modifiedParamValue, "NOT ");
+										}
+
+										//raise the alert, and save the attack string for the "Authentication Bypass" alert, if necessary
+										sqlInjectionAttack = modifiedParamValue;
+										bingo(Alert.RISK_HIGH, Alert.CONFIDENCE_MEDIUM, getName(), getDescription(),
+												null, //url
+												param, sqlInjectionAttack,
+												extraInfo, getSolution(), "", msg4);
+
+										sqlInjectionFoundForUrl = true;
+									}
+								}
+								//bale out if we were asked nicely
+								if (isStop()) { 
+									log.debug("Stopping the scan due to a user request");
+									return;
 								}
-
-								//raise the alert, and save the attack string for the "Authentication Bypass" alert, if necessary
-								sqlInjectionAttack = modifiedParamValue;
-								bingo(Alert.RISK_HIGH, Alert.CONFIDENCE_MEDIUM, getName(), getDescription(),
-										null, //url
-										param, sqlInjectionAttack,
-										extraInfo, getSolution(), "", msg4);
-
-								sqlInjectionFoundForUrl = true;
 							}
-						}
-						//bale out if we were asked nicely
-						if (isStop()) { 
-							log.debug("Stopping the scan due to a user request");
-							return;
+						} catch (Exception e){
```

Comment:
I believe this is to handle the case mentioned by @kingthorin about the overflow when adding/multiplying the int? If so, this needs changes since no exception is thrown in that case.